From 9446ef5232a076cc9cccd07b564922c69378db7e Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Mon, 29 Mar 2021 17:22:00 +0300
Subject: [PATCH] [SW-36011] gaudi: configure nic phy regs

Configure NIC phy registers according to nic enable
bitmask. This configuration is done on receiving
fit on secured transport (ELBI). The patch disables
or enables the port according to the bitmask.
A later patch shall protect this registers access.
In this way, NIC ports are enabled from FW only,
if secured transport (ELBI) writes the NIC
enable fit.

Change-Id: I1b98852b913f5045370d3f955fdc460f7102d35c
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/gaudi/common.c b/gaudi/common.c
index d844874..9c22025 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -17,6 +17,7 @@
 #include <pcie_dbi_regs.h>
 #include <cpu_timestamp_regs.h>
 #include <gic_regs.h>
+#include <nic0_phy_regs.h>
 
 #include <pcie_aux_masks.h>
 #include <pcie_dbi_masks.h>
@@ -466,9 +467,33 @@
 
 static u32 nic_en;
 
+#define NIC_PHY_RX_CFG_OFFSET	0x10
+#define NIC_PORT_EN		0
+#define NIC_PORT_DIS		2
+
+void enable_nic_regs(u32 nic_en)
+{
+	u32 port, val, idx, block_of;
+
+	for (port = 0; port < GAUDI_NIC_ENGS_NUM; port++) {
+		val = (nic_en & (1 << port)) ? NIC_PORT_EN : NIC_PORT_DIS;
+		idx = port >> 1;
+		block_of = idx * GAUDI_NICS_DIST;
+
+		if (port & 1) {
+			regmap_writel(val, mmNIC0_PHY_PHY_RX_CFG_2 + block_of);
+			regmap_writel(val, mmNIC0_PHY_PHY_RX_CFG_3 + block_of);
+		} else {
+			regmap_writel(val, mmNIC0_PHY_PHY_RX_CFG_0 + block_of);
+			regmap_writel(val, mmNIC0_PHY_PHY_RX_CFG_1 + block_of);
+		}
+	}
+}
+
 void preboot_nic_en(u32 nic_en_)
 {
 	nic_en = nic_en_;
+	enable_nic_regs(nic_en);
 	preboot_puts_hex("nic_en ", nic_en, true);
 	boot_info_set(HL_BOOT_INFO_NIC_EN, &nic_en, sizeof(nic_en));
 }
